report/04_implementation/01_flutter_application.md

```diff
@@ -6,23 +6,149 @@ web application) with a single code base.
+app.  
+While the provider package ensures the existence of only one service provider 
+instance, a critical issues arises when initializing all key parts. This is 
+mostly because of the asynchronous execution of code parts which have to be 
+synchronized. Solving this issue needed some sort of *Mutex* to be 
+implemented.  
+Unfortunately, *Dart* is kind of restrictive when it comes to using 
+rudimentary programming principles like *Mutexes*. Therefore, the `init` 
+method of the service provider takes a so called `Closure` which can be 
+described as nameless function. This function is invoked after all 
+initialization tasks have been completed.  
+The code example below explains this process in a more detailed manner.
+
+```dart
+// This class represents the actual app and the main widget.
+class SoTiredApp extends StatelessWidget {
+  const SoTiredApp({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    
+    ...
+    
+    // Register the Serviceprovider via the ChangeNotifierProvider provided 
+    // by the provider package.
+    return ChangeNotifierProvider(
+      create: (BuildContext context) => ServiceProvider(),
+      child: const SoTiredAppContent(),
+    );
+  }
+}
+
+...
+
+class SoTiredAppState extends State<SoTiredAppContent>
+    with WidgetsBindingObserver {
+  
+  ...
+
+  // This method is called immediately after every app start
+  @override
+  void initState() {
+    // call super method and instantiate widget observer
+
+    // Provider.of is part of the provider package and returns a registered
+    // provider. After that you can call every method this provider contains.
+    // The unnamed method the init function receives will set the variable 
+    // _doneInitializing to true which allows the typical Flutter build method 
+    // to ensure that every key has been initialized and returns the actual 
+    // app.
+    Provider.of<ServiceProvider>(context, listen: false)
+        .init(() => setState(() => _doneInitializing = true), path);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (_doneInitializing) {
+      return MaterialApp(...);
+    }
+  }
+
+  ...
+
+}
+
+...
+
+class ServiceProvider extends ChangeNotifier {
+  final ConfigManager _configManager = ConfigManager();
+  final DatabaseManager _databaseManager = DatabaseManager();
+  final Notifications _notifications = Notifications();
+
+  // ...
+  // Getter and Setter
+  // ...
+
+  Future<void> init(Function onDoneInitializing, <more arguments> ...) async {
+    // ...
+    // initialization tasks
+    // ...
+
+    onDoneInitializing(); // <- invoke this method after all init tasks have 
+                          // completed
+  }
+}
+```
+
+### Configuration Manager
+
+As mentioned in [Service Provider](service-provider) the configuration manager 
+is part of it. It also implements the Singleton design pattern principle to 
```

Comment:
part of what?
```suggestion
is part of the service provider. It also implements the Singleton design pattern principle to 
```

Reply:
Since this sentence already contains a reference to the service provider, I think it is clear that `it` is referring to the service provider. Therefore, I'd prefer omitting this suggestion as well 😅

Reply:
The only problem I had while reading was that I had to read twice what was meant. But yeah, I got it after a while.